pkg/upstream/cluster/loadbalancer.go

```diff
@@ -624,3 +625,146 @@ func (lb *reqRoundRobinLoadBalancer) IsExistsHosts(metadata api.MetadataMatchCri
 func (lb *reqRoundRobinLoadBalancer) HostNum(metadata api.MetadataMatchCriteria) int {
 	return lb.hosts.Size()
 }
+
+const defaultPeakEwmaDuration = 10 * time.Second
+
+type peakEwmaLoadBalancer struct {
+	*EdfLoadBalancer
+	rrLB types.LoadBalancer
+
+	choice          uint32
+	defaultDuration time.Duration
+}
+
+func newPeakEwmaLoadBalancer(info types.ClusterInfo, hosts types.HostSet) types.LoadBalancer {
+	lb := &peakEwmaLoadBalancer{
+		rrLB: rrFactory.newRoundRobinLoadBalancer(info, hosts),
+
+		choice:          defaultChoice,
+		defaultDuration: info.ConnectTimeout() + info.IdleTimeout(),
+	}
+
+	lb.EdfLoadBalancer = newEdfLoadBalancer(info, hosts, lb.unweightedChoose, lb.hostWeight)
+
+	if lb.defaultDuration == 0 {
+		lb.defaultDuration = defaultPeakEwmaDuration
+	}
+
+	return lb
+}
+
+func (lb *peakEwmaLoadBalancer) hostWeight(item WeightItem) float64 {
+	host, ok := item.(types.Host)
+	if !ok {
+		return 1.0
+	}
+
+	stats := host.HostStats()
+
+	duration := stats.UpstreamRequestDurationEWMA.Rate()
+	// None of the active requests returned, or the metrics is disabled
+	if duration == 0 {
+		duration = float64(lb.defaultDuration)
+	}
+
+	return float64(host.Weight()) / (duration * float64(stats.UpstreamRequestActive.Count()+1))
+}
+
+func (lb *peakEwmaLoadBalancer) unweightedChoose(context types.LoadBalancerContext) types.Host {
+	hs := lb.hosts
+
+	var candidate types.Host
+
+	// If `total` is less than or equal to `choice`, we can iterate over all elements directly.
+	if hs.Size() <= int(lb.choice) {
+		candidate = lb.iterateChoose()
+	} else {
+		candidate = lb.randomChoose()
+		if candidate == nil {
+			if log.DefaultLogger.GetLogLevel() >= log.WARN {
+				log.DefaultLogger.Warnf("[lb][PeakEwma] no host chosen after %d choice, fallback to RR", lb.choice)
+			}
+
+			return lb.rrLB.ChooseHost(context)
+		}
+	}
+
+	return candidate
+}
+
+func (lb *peakEwmaLoadBalancer) iterateChoose() types.Host {
+	total := lb.hosts.Size()
+
+	var candidate types.Host
+
+	// avoid always choosing the same host when the score is the same
+	lb.mutex.Lock()
+	idx := lb.rand.Intn(total)
+	lb.mutex.Unlock()
```

Comment:
IMHO, it is not so worth adding additional random generation & lock operations in the hot path.
Because when a host is chosen, its UpstreamRequestActive count will be increased, and its score is hardly the same the next time.

Reply:
> Because when a host is chosen, its UpstreamRequestActive count will be increased, and its score is hardly the same the next time.

There is a special case here, that is, if metrics are disabled (https://github.com/mosn/mosn/pull/2253#discussion_r1147359057), then if we start traversing from a fixed position, all requests will be skewed to the same host.

And, I don't think the cost of the lock here will be very high, in the iterative branch, it will only be executed once, while in the random version it will be executed at least `choice` times

Reply:
What a pity! Could you add a comment that this is for the special case?

Reply:
> What a pity! Could you add a comment that this is for the special case?

Completed 😺